torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h

```diff
@@ -0,0 +1,256 @@
+      output = IrBuilder::create<torch::jit::fuser::cuda::ComplexDouble>();
+    } else if (dtype_ == NvfDataType::Bool) {
+      output = IrBuilder::create<torch::jit::fuser::cuda::Bool>();
+    } else if (dtype_ == NvfDataType::Int) {
+      output = IrBuilder::create<torch::jit::fuser::cuda::Int>();
+    } else {
+      TORCH_CHECK(false, "Dtype is not supported:", dtype_);
+    }
+    fd.addInput(output);
+    fd.fusion_state.at(outputs.at(0)) = output;
+  }
+
+ private:
+  NvfDataType dtype_;
+};
+
+template <typename ExprType, typename ValueType>
+struct ConstantRecord : RecordFunctor {
+  ConstantRecord(std::vector<size_t> _outputs, ValueType val)
+      : RecordFunctor({}, std::move(_outputs)), value_(val) {}
+
+  void operator()(FusionDefinition& fd) final {
+    NvfVal* output = IrBuilder::create<ExprType>(value_);
+    fd.fusion_state.at(outputs.at(0)) = output;
+  }
+
+ private:
+  ValueType value_;
+};
+
+template <class OutputType>
+struct OutputRecord : RecordFunctor {
+  OutputRecord(std::vector<size_t> _args)
+      : RecordFunctor(std::move(_args), {}) {}
+
+  void operator()(FusionDefinition& fd) final {
+    auto input = fd.fusion_state.at(args.at(0));
+
+    // With C++17, this statement should be "if constexpr"
+    if (std::is_same<OutputType, NvfTensorView>::value) {
+      fd.addOutput(input->as<NvfTensorView>());
+    } else {
+      fd.addOutput(input);
+    }
+  }
+};
+
+struct BroadcastOpRecord : RecordFunctor {
+  BroadcastOpRecord(
+      std::vector<size_t> _args,
+      std::vector<size_t> _outputs,
+      std::vector<int64_t>& output_shape,
+      std::vector<int64_t>& broadcast_dims)
+      : RecordFunctor(std::move(_args), std::move(_outputs)),
+        output_shape_(std::move(output_shape)),
+        broadcast_dims_(std::move(broadcast_dims)) {}
+  void operator()(FusionDefinition& fd) final {
+    auto arg = fd.fusion_state.at(args.at(0))->as<TensorView>();
+
+    const auto arg_ndims = arg->domain()->noReductions().size();
+    TORCH_CHECK(
+        output_shape_.size() >= arg_ndims,
+        "The new shape is expected to be greater-then-or-equal to the input",
+        output_shape_.size(),
+        arg_ndims);
+    TORCH_CHECK(
+        arg_ndims == broadcast_dims_.size(),
+        "The broadcast dimensions should match the input dimensions.",
+        arg_ndims,
+        broadcast_dims_.size());
+
+    std::vector<bool> is_broadcast_dim(output_shape_.size(), true);
+    for (const auto idx : c10::irange(broadcast_dims_.size())) {
+      if (idx > 0) {
+        TORCH_CHECK(
+            broadcast_dims_[idx - 1] < broadcast_dims_[idx],
+            "Broadcast dimension is not greater than the previous value.");
+      }
+      TORCH_CHECK(
+          broadcast_dims_[idx] < static_cast<int>(output_shape_.size()),
+          "Invalid broadcast_dims value.");
+      is_broadcast_dim.at(broadcast_dims_[idx]) = false;
+    }
+
+    auto output = torch::jit::fuser::cuda::broadcast(arg, is_broadcast_dim);
```

Comment:
hint: we should consider supporting `expand` as well. Looking at this: https://github.com/pytorch/pytorch/blob/d72040335b58e03115ad7d6e43e9099fd3e9e1c4/torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp#L617-L679

Reply:
I haven't merged in a week, from upstream.  It looks like I need to.

Reply:
This change is not in upstream.  Therefore, I suggest we don't add it pre-maturely to create more merge difficulties.  I have been exclusively making edits on the frontend relative to upstream since Primtorch has a dependency on it.